### PR TITLE
Adding bindings/tflite/ to expose IREE through the tflite C API

### LINKED ISSUE #3952: Add/find/port/etc TOSA->Linalg lowerings for ops needed for the target model.
This relates to the linalg on tensors work in that we may be able to bypass all of the legacy linalg on buffers codegen parts if we are able to get things working in parallel. If not, copy/pasting some of our HLO->linalg-on-buffers conversion patterns is acceptable - it's all dead code anyway. The goal is to not have a production-ready and exhaustively ported set of patterns but to literally have some of the [8 non-trivial ops that TFLite GPU supports](https://www.tensorflow.org/lite/performance/gpu_advanced#supported_ops).

### PR DESCRIPTION
**Pull request by benvanik**
Somewhat surprisingly, this seems to work. Some gotchas and TODOs
remaining as tracked in https://github.com/google/iree/projects/17,
but it can run a hello-world add module hand-lowered from a tflite
flatbuffer to an IREE input module. Future work in #3974/#3978/#3972
will automate this lowering.

Fixes #3952.

### CODE REVIEW

**Review by ScottTodd — APPROVED** [review]

**Review by not-jenni — APPROVED** [review]

File: bindings/tflite/shim.c [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+#include "bindings/tflite/shim.h"
+
+// NOTE: we pull in our own copy here in case the tflite API changes upstream.
+#define TFL_COMPILE_LIBRARY 1
+#include "bindings/tflite/include/tensorflow/lite/c/c_api.h"
+#include "bindings/tflite/include/tensorflow/lite/c/c_api_experimental.h"
+
+TFL_CAPI_EXPORT extern const char* TfLiteVersion(void) { return "👻"; }
```
[the remark this answers is not in the document]
Comment (benvanik) [reply bindings/tflite/shim.c:22]: This made me so happy. It's just as good as a fourcc/magic identifier, as in utf8 string form it's 4 bytes: http://www.ltg.ed.ac.uk/~richard/utf-8.cgi?input=1F47B&mode=hex - going to use it everywhere we need magic :)

File: bindings/tflite/interpreter.c [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+      p, iree_vm_list_storage_size(&buffer_view_type_def, model->output_count));
+  IREE_RETURN_IF_ERROR(
+      iree_vm_list_initialize(output_list_storage, &buffer_view_type_def,
+                              model->output_count, &interpreter->output_list));
+  p += output_list_storage.data_length;
+
+  interpreter->input_tensors = (TfLiteTensor*)p;
+  p += sizeof(TfLiteTensor) * model->input_count;
+  interpreter->output_tensors = (TfLiteTensor*)p;
+  // p += sizeof(TfLiteTensor) * model->output_count;
```
[the remark this answers is not in the document]
Comment (benvanik) [reply bindings/tflite/interpreter.c:292]: yeah - I feel like one of our bots had a warning-as-error of it being unused (sigh), but it's good documentation (adding another field would be at position p).

File: bindings/tflite/include/tensorflow/lite/c/c_api.h [hunk 3]
```diff
@@ -0,0 +1,255 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply bindings/tflite/include/tensorflow/lite/c/c_api.h:1]: good point - they do have some modifications guarded with `IREE_BINDINGS_TFLITE_INCLUDE_UNSUPPORTED_APIS`, but I'll add a readme with the original hash and notes about the modifications.

File: bindings/tflite/shim.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+
+#ifndef IREE_BINDINGS_TFLITE_SHIM_H_
+#define IREE_BINDINGS_TFLITE_SHIM_H_
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/vm/api.h"
+
+// NOTE: we pull in our own copy here in case the tflite API changes upstream.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply bindings/tflite/shim.h:22]: mentioned in new include/README.md next to the copied files

File: bindings/tflite/README.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
+# IREE TFLite C API Compatibility Shim
+
+Provides a (mostly) tflite-compatible API that allows loading compiled IREE
+modules, managing tensors, and invoking functions with the same conventions as
+[TFLite's C API](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/c).
+
+The intent is not to have a faithful reproduction of all tflite features as
+IREE obviates the need for many of them (such as builtin ops). This is meant as
+a way for applications that may currently be using the tflite API to quickly
+onboard with IREE.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply bindings/tflite/README.md:10]: added disclaimer and also linked to the project

File: bindings/tflite/README.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+Though it's possible to use multiple `TfLiteInterpreter` instances in the same
+process in the real tflite it is strongly discouraged: each interpreter will create its own thread and memory pools and device handles to accelerators and
+assume it owns all resources exclusively. The experimental external contexts
+API is present to try to allow for something better than that and IREE would be
+able to make use of it to the extent the feature allows.
+
+But IREE is designed to fully support large constellations of models all running
+concurrently and passing data between both each other and the application
+efficiently pipelined across cross-device and cross-process. Though external
```
Comment (antiagainst) [inline bindings/tflite/README.md:158]: remove across?

File: bindings/tflite/interpreter.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+#define IREE_BINDINGS_TFLITE_INTERPRETER_H_
+
+#include "bindings/tflite/model.h"
+#include "bindings/tflite/options.h"
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/vm/api.h"
+
+// NOTE: we pull in our own copy here in case the tflite API changes upstream.
+#define TFL_COMPILE_LIBRARY 1
```
Comment (antiagainst) [inline bindings/tflite/interpreter.h:25]: Does this need to be undefed?

**Review by antiagainst — COMMENTED** [review]
Fantastic!

File: bindings/tflite/interpreter.h (hunk 7 above)
Comment (benvanik) [reply bindings/tflite/interpreter.h:25]: nah this is all internal code (not included by user files)

**Pull request merged**